lib/linter/linter.js

```diff
@@ -1537,102 +1556,86 @@ class Linter {
         const filename = options.filename || "<input>";
         const filenameToExpose = normalizeFilename(filename);
         const physicalFilename = options.physicalFilename || filenameToExpose;
         const text = ensureText(textOrSourceCode);
+        const file = new VFile(filenameToExpose, text, {
+            physicalPath: physicalFilename
+        });
+
         const preprocess = options.preprocess || (rawText => [rawText]);
         const postprocess = options.postprocess || (messagesList => messagesList.flat());
+
+        const processorService = new ProcessorService();
+        const preprocessResult = processorService.preprocessSync(file, {
+            processor: {
+                preprocess,
+                postprocess
+            }
+        });
+
+        if (!preprocessResult.ok) {
+            return preprocessResult.errors;
+        }
+
         const filterCodeBlock =
             options.filterCodeBlock ||
             (blockFilename => blockFilename.endsWith(".js"));
         const originalExtname = path.extname(filename);
+        const { files } = preprocessResult;
 
-        let blocks;
-
-        try {
-            blocks = preprocess(text, filenameToExpose);
-        } catch (ex) {
-
-            // If the message includes a leading line number, strip it:
-            const message = `Preprocessing error: ${ex.message.replace(/^line \d+:/iu, "").trim()}`;
-
-            debug("%s\n%s", message, ex.stack);
-
-            return [
-                {
-                    ruleId: null,
-                    fatal: true,
-                    severity: 2,
-                    message,
-                    line: ex.lineNumber,
-                    column: ex.column,
-                    nodeType: null
-                }
-            ];
-        }
-
-        const messageLists = blocks.map((block, i) => {
-            debug("A code block was found: %o", block.filename || "(unnamed)");
+        const messageLists = files.map(block => {
+            debug("A code block was found: %o", block.path || "(unnamed)");
 
             // Keep the legacy behavior.
             if (typeof block === "string") {
                 return this._verifyWithFlatConfigArrayAndWithoutProcessors(block, config, options);
             }
 
-            const blockText = block.text;
-            const blockName = path.join(filename, `${i}_${block.filename}`);
-
             // Skip this block if filtered.
-            if (!filterCodeBlock(blockName, blockText)) {
+            if (!filterCodeBlock(block.path, block.body)) {
                 debug("This code block was skipped.");
                 return [];
             }
 
             // Resolve configuration again if the file content or extension was changed.
-            if (configForRecursive && (text !== blockText || path.extname(blockName) !== originalExtname)) {
+            if (configForRecursive && (text !== block.body || path.extname(block.path) !== originalExtname)) {
```

Comment:
The example from https://github.com/eslint/eslint/pull/18818#discussion_r1743608645 still doesn't work, because `text !== block.body` is `true` when file has BOM so it enters this code path and lints `block.body` which doesn't include BOM.